source/common/http/codec_client.cc

```diff
@@ -128,8 +128,25 @@ void CodecClient::responsePreDecodeComplete(ActiveRequest& request) {
   if (codec_client_callbacks_) {
     codec_client_callbacks_->onStreamPreDecodeComplete();
   }
-  deleteRequest(request);
 
+  request.decode_complete_ = true;
+  if (request.encode_complete_) {
+    completeRequest(request);
+  } else {
+    ENVOY_CONN_LOG(debug, "waiting for encode to complete", *connection_);
+  }
+}
+
+void CodecClient::onRequestStreamEnd(ActiveRequest& request) {
```

Comment:
Envoy tends towards Decode/Encode instead of Request/Response :-(

Reply:
Thanks, will update. It's quite confusing since codec is used upstream.